activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -732,11 +732,50 @@ def raw_execute(sql, name, async: false, allow_retry: false, uses_transaction: t
           log(sql, name, async: async) do
             with_raw_connection(allow_retry: allow_retry, uses_transaction: uses_transaction) do |conn|
               sync_timezone_changes(conn)
-              conn.query(sql)
+              result = conn.query(sql)
+              handle_warnings(sql)
+              result
             end
           end
         end
 
+        def handle_warnings(sql)
+          return if ActiveRecord.db_warnings_action == :ignore || @raw_connection.warning_count == 0
+
+          result = @raw_connection.query("SHOW WARNINGS")
+          result.each do |level, code, message|
+            warning = SqlWarning.new(message, code, level, sql)
+            next if warning_ignored?(warning)
+
+            case ActiveRecord.db_warnings_action
+            when :log
+              logger.warn("[#{warning.class}] #{warning.message} (#{warning.code})")
+            when :raise
+              raise warning
+            when Proc
+              ActiveRecord.db_warnings_action.call(warning)
+            end
```

Comment:
Rather than a case, it might make sense to convert the symbol into a callable during boot.

e.g.

```
module ActiveRecord
  class << self
    attr_reader :db_warnings_action 
    def db_warnings_action=(action)
      @db_warnings_action = case action
      when :log
        ->(warning) { logger.warn("[#{warning.class}] #{warning.message} (#{warning.code})") }
      when :raise
        ->(warning) { raise warning }
      # ...snip
      end
    end
  end
end
```

Reply:
Also would be nice to have a `:report` option that calls `Rails.error.report`.